### PR TITLE
Allow linking UCC in nvfuser build

### PR DESCRIPTION
**Pull request by xwang233**
After this PR, nvfuser build may link with UCC.

For submodule build, no change is needed. If the pytorch build has UCC enabled, nvfuser build will link with UCC, too.

For standalone top level build, UCC is **not** enabled in nvfuser by default. User will need to specify `--build-with-ucc` with setup.py to enable build with UCC. This only happens if pytorch is built with UCC as well.

This PR also fixed a minor issue in C++ definitions of `ProcessGroupUCC` from pytorch.

Related #483

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -219,6 +219,17 @@ target_link_libraries(${NVFUSER_CODEGEN} PRIVATE stdc++fs)
...
+if (NOT UCC_FOUND)
+  # This probably means a "standalone" build, where UCC_FOUND is not inherited from pytorch's CMake configuration.
+  # It's required to set env vars UCC_DIR, UCX_DIR, UCC_HOME, UCX_HOME for CMake's find UCC to work.
+  find_package(UCC REQUIRED)
+  find_package(UCX REQUIRED)
+endif()
+add_library(__nvfuser_ucc INTERFACE)
+target_link_libraries(__nvfuser_ucc INTERFACE ucx::ucs ucx::ucp ucc::ucc)
+target_include_directories(__nvfuser_ucc INTERFACE ${UCC_INCLUDE_DIRS})
+target_link_libraries(${NVFUSER_CODEGEN} PRIVATE __nvfuser_ucc)
```
Comment (zasdfgbnm) [inline CMakeLists.txt:231]: I think you would need to check if PyTorch is built with UCC, and only do so if PyTorch is.

File: CMakeLists.txt [hunk 2]
```diff
@@ -219,6 +219,17 @@ target_link_libraries(${NVFUSER_CODEGEN} PRIVATE stdc++fs)
 target_link_libraries(${NVFUSER_CODEGEN} PRIVATE ${CUDA_NVRTC_LIB} ${LIBNVTOOLSEXT})
 target_include_directories(${NVFUSER_CODEGEN} PRIVATE ${CUDA_INCLUDE_DIRS})
 
+if (NOT UCC_FOUND)
+  # This probably means a "standalone" build, where UCC_FOUND is not inherited from pytorch's CMake configuration.
+  # It's required to set env vars UCC_DIR, UCX_DIR, UCC_HOME, UCX_HOME for CMake's find UCC to work.
+  find_package(UCC REQUIRED)
+  find_package(UCX REQUIRED)
```
Comment (jjsjann123) [inline CMakeLists.txt:226]: Looks like CI node doesn't have ucc installed.

Looking at Xiang's PR, there seems to be a script doing that already: https://github.com/pytorch/pytorch/commit/cda210e23b6bb42658e7e4f70778f9db7a900645#diff-a45bde6f6abdaabff96d1742357b0817fd3e3e98dd7e534a744fb955690d2ba3

.circleci/docker/common/install_ucc.sh

File: CMakeLists.txt (hunk 1 above)
Comment (jjsjann123) [reply CMakeLists.txt:231]: dumb question, is pytorch exposing USE_UCC in config file? How are we checking that on a pip installed torch?

Comment (zasdfgbnm) [reply CMakeLists.txt:231]: On the python level, yes: https://github.com/pytorch/pytorch/blob/1a6619a830ad0fa7e364f1a4b178daf3d78b6e4c/test/run_test.py#L408

On CMake level, I don't know...

File: CMakeLists.txt (hunk 2 above)
Comment (xwang233) [reply CMakeLists.txt:226]: UCC is installed by default in NGC cuda container

Comment (jjsjann123) [reply CMakeLists.txt:226]: I meant the standalone build ci using github nodes: https://github.com/NVIDIA/Fuser/blob/41f4dce9c8e97f811ba5704b7a69c56061ab8672/.github/workflows/lint.yml#L13-L29

**Comment by xwang233** [comment]
I think an important question is, do we want nvfuser to link/include ucc:
- directly
- from ucc exposed by pytorch

**Comment by samnordmann** [comment]
> I think an important question is, do we want nvfuser to link/include ucc:
> 
> * directly
> * from ucc exposed by pytorch
> 
> cc @ptrblck

For now, for multi device support, pytorch's ProcessGroup API is enough, but in the future, we might want to bypass it

**Comment by xwang233** [comment]
!build

**Comment by xwang233** [comment]
The triggered CI build is using a no-UCC pytorch build.

Build & test with a UCC-enabled pytorch build is here https://nv/e2G/8952870

looks good so far

File: csrc/multidevice/communicator.cpp [hunk 3]
```diff
@@ -14,7 +14,7 @@
 #ifdef USE_C10D_NCCL
 #include <torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp>
 #endif
-#ifdef USE_C10D_UCC
+#if defined(USE_C10D_UCC) && defined(NVFUSER_BUILD_WITH_UCC)
```
Comment (jjsjann123) [inline csrc/multidevice/communicator.cpp:17]: This looks ugly. But I understand that we are just doing the minimal to get CI green so it's fine for now.

File: setup.py [hunk 4]
```diff
@@ -80,6 +84,9 @@
     if arg == "--no-ninja":
         NO_NINJA = True
         continue
+    if arg == "--build-with-ucc":
```
Comment (jjsjann123) [inline setup.py:87]: I think we also needed a cmake_prefix_path to actually find where the UCC libraries are? not sure if that part can be done automatically. i.e. do they have some python module/utils to query where the library is?

**Review by jjsjann123 — APPROVED** [review]
stamp for green CI.

But frankly, we need to do a better job isolating `optional` dependencies.

File: csrc/multidevice/communicator.cpp (hunk 3 above)
Comment (xwang233) [reply csrc/multidevice/communicator.cpp:17]: 😟

**Pull request merged**